source/slang/slang-ir-lower-buffer-element-type.cpp

```diff
@@ -914,33 +1143,552 @@ struct LoweredElementTypeContext
+                                    logicalType,
+                                    storageGEP,
+                                    castInst->getBufferType());
+                                user->replaceUsesWith(castOfGEP);
+                                user->removeAndDeallocate();
+                                if (auto castStorage = as<IRCastStorageToLogical>(castOfGEP))
+                                    castInstWorkList.add(castStorage);
+                                break;
+                            }
+                        case kIROp_Call:
+                            {
+                                // call(f, castStorageToLogical(x)) ==> call(f', x)
+                                //
+                                // If we see a call that takes a logical typed pointer, we will
+                                // specialize the callee to take a storage typed pointer instead,
+                                // and push the cast to inside the callee.
+                                // We will process calls after other gep insts, so for now just add
+                                // it into a separate worklist.
+                                if (castInst->getOp() == kIROp_CastStorageToLogical)
+                                {
+                                    callWorkListSet.add((IRCall*)user);
+                                }
+                                break;
+                            }
+                        case kIROp_Load:
+                        case kIROp_StructuredBufferLoad:
+                        case kIROp_RWStructuredBufferLoad:
+                        case kIROp_StructuredBufferLoadStatus:
+                        case kIROp_RWStructuredBufferLoadStatus:
+                        case kIROp_StructuredBufferConsume:
+                            {
+                                // If we see a load(CastStorageToLogical(storageAddr)),
+                                // then based on what `storageAddr` is, we will push down
+                                // the cast differently.
+                                // - If `storageAddr` is already a tempVar that we introduced to
+                                //   hold the value of a buffer resource load, we can simply
+                                //   convert this into `CastStorageToLogicalDeref(storageAddr)`.
+                                // - Otherwise, if `storageAddr` is a buffer location, we will
+                                //   create a temp var to hold the result of the memory load,
+                                //   Then we create a `CastStorageToLogicalDeref(tempVar)`
+                                //   structure and use it to replace `user`.
+                                // Note that it is important to introduce a temp var and preserve
+                                // the buffer load operation, so we are not changing the memory
+                                // semantics of the original program.
```

Comment:
I think this comment is related to some of my earlier concerns: at the point where a load happens, this pass introduces a temp (using the storage type), to ensure that the contents of the memory are observed at the right point.